drivers/usb/device/usb_dc_stm.c

```diff
@@ -0,0 +1,644 @@
+#define EP_IS_OUT(ep) (((ep) & USB_EP_DIR_MASK) == USB_EP_DIR_OUT)
+
+/* Endpoint state */
+
+struct usb_dc_stm_ep_state {
+	u16_t ep_mps;	/** Endpoint max packet size */
+	u8_t ep_type;	/** Endpoint type (STM32 HAL enum) */
+	usb_dc_ep_callback cb;	/** Endpoint callback function */
+	u8_t ep_stalled;	/** Endpoint stall flag */
+	u32_t read_count;	/** Number of bytes in read buffer  */
+	u32_t read_offset;	/** Current offset in read buffer */
+};
+
+/* Driver state */
+
+struct usb_dc_stm_state {
+	PCD_HandleTypeDef pcd;	/* Storage for the HAL_PCD api */
+	usb_dc_status_callback status_cb; /* Status callback */
+	struct usb_dc_stm_ep_state out_ep_state[CONFIG_USB_DC_STM_EP_NUM];
+	struct usb_dc_stm_ep_state in_ep_state[CONFIG_USB_DC_STM_EP_NUM];
+	u8_t ep_buf[CONFIG_USB_DC_STM_EP_NUM][USB_OTG_FS_MAX_PACKET_SIZE];
+};
+
+static struct usb_dc_stm_state usb_dc_stm_state;
+
+/* Internal functions */
+
+static struct usb_dc_stm_ep_state *usb_dc_stm_get_ep_state(u8_t ep)
+{
+	struct usb_dc_stm_ep_state *ep_state_base;
+
+	if (EP_IDX(ep) >= CONFIG_USB_DC_STM_EP_NUM) {
+		return NULL;
+	}
+
+	if (EP_IS_OUT(ep)) {
+		ep_state_base = usb_dc_stm_state.out_ep_state;
+	} else {
+		ep_state_base = usb_dc_stm_state.in_ep_state;
+	}
+
+	return ep_state_base + EP_IDX(ep);
+}
+
+static void usb_dc_stm_isr(void *arg)
+{
+	HAL_PCD_IRQHandler(&usb_dc_stm_state.pcd);
+}
+
+static int usb_dc_stm_clock_enable(void)
+{
+	struct device *clk = device_get_binding(STM32_CLOCK_CONTROL_NAME);
+	struct stm32_pclken pclken = {
+		.bus = STM32_CLOCK_BUS_AHB2,
+		.enr = LL_AHB2_GRP1_PERIPH_OTGFS,
+	};
+
+	clock_control_on(clk, (clock_control_subsys_t *)&pclken);
+
+	return 0;
+}
+
+static int usb_dc_stm_init(void)
+{
+	HAL_StatusTypeDef status;
+	unsigned int i;
+
+	usb_dc_stm_state.pcd.Instance = USB_OTG_FS;
+	usb_dc_stm_state.pcd.Init.dev_endpoints = 3;
+	usb_dc_stm_state.pcd.Init.speed = USB_OTG_SPEED_FULL;
+	usb_dc_stm_state.pcd.Init.phy_itface = PCD_PHY_EMBEDDED;
+	usb_dc_stm_state.pcd.Init.ep0_mps = EP0_MPS;
+	usb_dc_stm_state.pcd.Init.dma_enable = DISABLE;
+
+#if CONFIG_USB_DC_STM_VBUS_SENSING
+	usb_dc_stm_state.pcd.Init.vbus_sensing_enable = ENABLE;
+#else
+	usb_dc_stm_state.pcd.Init.vbus_sensing_enable = DISABLE;
+#endif
+
+	SYS_LOG_DBG("HAL_PCD_Init");
+	status = HAL_PCD_Init(&usb_dc_stm_state.pcd);
+	if (status != HAL_OK) {
+		SYS_LOG_ERR("PCD_Init failed, %d", (int)status);
+		return -EIO;
+	}
+
+	SYS_LOG_DBG("HAL_PCD_Start");
+	status = HAL_PCD_Start(&usb_dc_stm_state.pcd);
+	if (status != HAL_OK) {
+		SYS_LOG_ERR("PCD_Start failed, %d", (int)status);
+		return -EIO;
+	}
+
+	usb_dc_stm_state.out_ep_state[EP0_IDX].ep_mps = EP0_MPS;
+	usb_dc_stm_state.out_ep_state[EP0_IDX].ep_type = EP_TYPE_CTRL;
+	usb_dc_stm_state.in_ep_state[EP0_IDX].ep_mps = EP0_MPS;
+	usb_dc_stm_state.in_ep_state[EP0_IDX].ep_type = EP_TYPE_CTRL;
+
+	HAL_PCDEx_SetRxFiFo(&usb_dc_stm_state.pcd, FIFO_SIZE_IN_WORDS);
+	for (i = 0; i < CONFIG_USB_DC_STM_EP_NUM; i++) {
```

Comment:
Here again, what gives CONFIG_USB_DC_STM_EP_NUM ?

Reply:
I updated it with the number of FIFOs.